app/webpacker/controllers/variant_controller.js

```diff
@@ -0,0 +1,64 @@
+import { Controller } from "stimulus";
+
+// Dynamically update related variant fields
+//
+export default class VariantController extends Controller {
+  static targets = ["display_as", "unit_to_display"]; // todo: variant_unit, variant_unit_scale
+
+  connect() {
+    // Assuming these will be available on the variant soon, just a quick hack to find the product fields:
+    const product = this.element.closest("[data-record-id]");
+    this.variantUnit = product.querySelector('[name$="[variant_unit]"]');
+    this.variantUnitScale = product.querySelector('[name$="[variant_unit_scale]"]');
+    this.variantUnitName = product.querySelector('[name$="[variant_unit_name]"]');
+
+    this.unitValueWithDescription = this.element.querySelector(
+      '[name$="[unit_value_with_description]"]',
+    );
+    this.displayAs = this.element.querySelector('[name$="[display_as]"]');
+    this.unitToDisplay = this.element.querySelector('[name$="[unit_to_display]"]');
+
+    // on unit changed; update display_as:placeholder and unit_to_display
+    [this.variantUnit, this.variantUnitScale, this.variantUnitName].forEach((element) => {
+      element.addEventListener("change", this.#unitChanged.bind(this), { passive: true });
+    });
```

Comment:
Hang on, this is where it seem complicated. The variant unit is actually a hidden field, right? So is the product controller updating first and then the variant controller is working on those changes? I guess we need that product controller at the moment to supply the right values to Rails. And since we have it, we may as well use it here. :exploding_head: 
I think that the product refactor will make this easier.

Reply:
Yeah I'll admit this is pretty strange. I probably could have done it simpler, but after Product Refactor we can make it much simpler still (we shouldn't even need those hidden fields), so would wait until then to refactor.